In [1]:
from scipy import stats
import numpy as np

### Задание

Подкинули монету N раз. Кол-во случаев, когда выпал орёл, на 10% больше, чем кол-во случаев, когда выпала решка. При каком N мы можем сказать, что монета «нечестная» (орёл и решка выпадают с разной вероятностью)?


Выборка представляет из себя - бинарный вектор длины N, где 1 - выпал орёл, 0 - выпала решка. 

H0 - нулевая гипотеза - монета честная, и вероятность выпадения орла и решки одинакова и равна 0.5

H1 - альтернативная гипотеза - монета нечестная. 

Пусть x - выпала решка, тогда выпадение орла = x + 0.1*x = 1.1x

Общее количество событий = количество выпадений решки + количество выпадений орла, т.е. N = x+1.1x = 2.1x

Таким образом, выпадение решки = N/2.1, а выпадение орла = 1.1 * N/2.1 = 0.524*N

Предположим, что верна нулевая гипотеза. В таком случае успех и неудача происходят равновероятно. 

In [2]:
for n in np.arange(2,2000):
    F_H0 = stats.binom(n,0.5)
    p = stats.binom_test(0.524*n,n,0.5, alternative='two-sided')
    if p < 0.05: #проверка на уровне значимости 0.05
        print(f'Количество экспериментов = {n}, значение p-value = {p}')
        break

Количество экспериментов = 1731, значение p-value = 0.04870361520515429


Вывод - при N = 1731 мы можем отклонить нулевую гипотезу на уровне значимости 0.05 и сказать, что монеты нечестные. 